Add dual-core support for cy8c624abzi_s2d44 SOC and CY8CKIT-062S2-AI board

alexskp opened the pull request:
This PR adds dual-core support for the cy8c624abzi_s2d44 SOC and CY8CKIT-062S2-AI board,
enabling both Cortex-M0+ and Cortex-M4 cores to run Zephyr simultaneously.

## Changes

### SOC Support (cy8c624abzi_s2d44)
- Add dual-core SOC initialization for CY8C624ABZI-S2D44
- Implement memory partitioning (M0: 128KB, M4: 892KB, Shared: 4KB)
- Add interrupts support for M0 core
- Reimplement Cy_SysInt_Init function using Zephyr dynamic interrupts
- Configure M0 NVIC channels (M0 has only 8 NVIC inputs, connects interrupts to channels 0-7)
- Update SOC and driver code for dual-core operation

### Board Support (CY8CKIT-062S2-AI)
- Add dual-core support
- Add board variants with qualifiers:
  * `cy8ckit_062s2_ai/cy8c624abzi_s2d44/m4` - M4-only (default)
  * `cy8ckit_062s2_ai/cy8c624abzi_s2d44/m0` - M0-only
  * `cy8ckit_062s2_ai/cy8c624abzi_s2d44/m4_dual` - M4 in dual-core mode
- Add device tree files for M0 and M4 cores with proper memory partitioning
- Configure device tree for dual-core setup with separate flash and SRAM regions
- Update board documentation with variant information, memory partitioning details, and build examples
- Add sample application demonstrating dual-core operation using sysbuild
- Include OpenOCD debug configuration

## Testing
- Dual-core sample application runs on both cores
- M0+ handles LED0, M4 handles LED1
- Proper memory isolation between cores
- Interrupt handling working correctly on M0 core

Partially addresses #97807

alexskp committed:
boards: infineon: Add dual-core board support for CY8CKIT-062S2-AI-DUAL

- Add M0+ and M4 board configurations with proper memory partitioning
- Add board metadata and Kconfig definitions
- Add documentation and OpenOCD support
- Update existing board files to use M4-specific naming

Partially addresses #97807

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
soc: infineon: Add dual-core SOC support for CY8C624ABZI-S2D44

- Add M0-specific DTSI file for dual-core support
- Update SOC configuration for dual-core interrupt handling
- Update GPIO driver for M0 compatibility

Partially addresses #97807

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
samples: boards: Add dual-core sample for CY8CKIT-062S2-AI-DUAL

- Add sample application demonstrating dual-core coordination
- Include M0+ and M4 application code
- Add sysbuild configuration for dual-core builds
- Add sample documentation and README

Partially addresses #97807

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

nordicjm commented on boards/infineon/cy8ckit_062s2_ai_dual/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+
+# Select HCI components
+config UART
+	bool
```
do not redeclare existing variables

alexskp replied:
I'm following same style from another infineon board. This file copied from cy8cproto_062_4343w board

nordicjm replied:
then the other board is wrong, fix it

nordicjm commented on boards/infineon/cy8ckit_062s2_ai_dual/Kconfig.defconfig:
```diff
@@ -0,0 +1,47 @@
+	default 15000 if WIFI
+	default 4096
+
+endif   # BOARD_CY8CKIT_062S2_AI
```
lots of space before `#`, make it 1

nordicjm commented on soc/infineon/cat1a/common/soc.c:
```diff
@@ -14,6 +14,82 @@
+	return ((IRQn_Type)tempReg);
+}
+
+cy_en_intr_t Cy_SysInt_GetInterruptActive(IRQn_Type IRQn)
```
zephyr is snake case not camel case

nordicjm commented on samples/boards/infineon/cy8ckit_062s2_ai_dual_dual_core/CMakeLists.txt:
```diff
@@ -0,0 +1,7 @@
+cmake_minimum_required(VERSION 3.20.0)
+
+find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
```
we have a multi core sample already, why do we need another one that is board specific? Why can't you instead get yours working with the existing one https://github.com/zephyrproject-rtos/zephyr/tree/main/samples/sysbuild/hello_world

nordicjm replied:
That sample does the same for various different types of hardware, your hardware should be updated to work the same and use the same base sample

alexskp committed:
boards: infineon: Add dual-core support for CY8CKIT-062S2-AI

Refactor dual-core support from separate board to board variants
within cy8ckit_062s2_ai. This allows the same board to support
M4-only, M0-only, and dual-core configurations using board
qualifiers.

Changes:
- Add board variants: m0, m4, and m4_dual qualifiers
- Add M0 and M4 DTS files with proper memory partitioning:
  * M0+: 512KB flash, 128KB SRAM
  * M4: 1536KB flash, 892KB SRAM
  * Shared: 4KB SRAM for inter-core communication
- Update SOC configuration to declare dual-core CPU clusters
- Move dual-core sample from cy8ckit_062s2_ai_dual_dual_core to
  cy8ckit_062s2_ai_dual_core
- Update board documentation with variant information and
  memory partitioning details
- Configure clock control: M0+ handles clock configuration
  in dual-core mode to avoid hardware conflicts

The board can now be used with:
- cy8ckit_062s2_ai/cy8c624abzi_s2d44/m4 (M4-only, default)
- cy8ckit_062s2_ai/cy8c624abzi_s2d44/m0 (M0-only)
- cy8ckit_062s2_ai/cy8c624abzi_s2d44/m0 + m4_dual (dual-core)

Partially addresses #97807

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
boards: infineon: Fix build after upstream binding rename

Update device tree compatible strings from 'infineon,cat1-uart' to
'infineon,uart' to match upstream commit 0a06f5b that removed the
'cat1' prefix from Infineon bindings.

Also enable USE_INFINEON_LEGACY_HAL for PSOC6_M0 and PSOC6_M4 SOC
families to ensure mtb-hal-cat1 module is included, which is required
for clock control driver (cyhal_clock.h).

Fixes build errors after merging upstream/main.

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
samples: sysbuild: Add Infineon dual-core support to hello_world sample

Add support for CY8CKIT-062S2-AI dual-core configuration (M0+M4) to the
existing hello_world sysbuild sample, replacing the custom dual-core
sample application.

Changes:
- Add Kconfig support to automatically select M4 dual board when
  building for M0 core (SOC_CY8C624ABZI_S2D44_M0)
- Add test entry in sample.yaml for cy8ckit_062s2_ai/cy8c624abzi_s2d44/m0
- Add M4 core startup code in soc.c for M0 builds, calculating M4 flash
  base address from M0 flash region (M0 flash base + M0 flash size)
- Update other sample.yaml files to use board qualifier format for
  cy8cproto_062_4343w board

The hello_world sample now demonstrates dual-core sysbuild for Infineon
boards, with M0 building the main application and M4 building the
remote application automatically via sysbuild.

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
boards: infineon: Fix cy8ckit_062s2_ai documentation formatting

Fix RST formatting issues:
- Fix nested list indentation under Memory partitioning
- Change :zephyr-sample: to :zephyr:code-sample: (correct role)
- Update sample references to sysbuild.hello_world
- Update build command to use new sample path

Fixes documentation build errors.

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
boards: cy8ckit_062s2_ai: fix compliance check errors

Fix DTS formatting and documentation code-sample reference.

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

alexskp committed:
dts: infineon: fix interrupt priority define order

Move PSOC6_02_DEFAULT_IRQ_PRIORITY definition before the base DTSI
include to ensure correct interrupt priority configuration for both
M0+ and M4 cores.

Also update test configurations:
- Remove old board overlay files (replaced with core-specific ones)
- Add M0 platform to test allowlists where appropriate
- Exclude M0 platform from UART async API tests that require M4

Signed-off-by: Oleksandr Skopets <skopets.sasha@gmail.com>

teburd commented:
Please do fix the conflicts, we are working quite a lot to enable various peripherals and parts. If you can rebase this and fix the issues I'll be sure to look at it. If I miss it for too long reach out on discord my handle is teburd.

JarmouniA commented on drivers/gpio/gpio_infineon.c:
```diff
@@ -366,6 +366,14 @@ static void gpio_psoc4_register_irq(const struct device *dev)
 #define ENABLE_INT(n)
 
+#elif (CONFIG_SOC_FAMILY_PSOC6_M0)
+#define INTR_PRIORITY(n) .intr_priority = DT_INST_IRQ_BY_IDX(n, 0, priority),
```
Why is INTR_PRIORITY being duplicated!?

JarmouniA commented on dts/arm/infineon/cat1a/mpns/CY8C624ABZI_S2D44.dtsi:
```diff
@@ -8,6 +8,8 @@
 #include <arm/armv7-m.dtsi>
 #include "../psoc6_02/psoc6_02.124-bga.dtsi"
 
+#define PSOC6_02_DEFAULT_IRQ_PRIORITY 6
```
Should go into a dts-bindings header.

JarmouniA reviewed (changes requested):
No fix up commits please, correct things in the commit introducing them directly.
